gh-110481: Implement biased reference counting

colesbury opened the pull request:
CPython's current reference counting implementation would not be thread-safe without the GIL. This implements biased reference counting in `--disable-gil` builds, which is thread-safe and has lower execution overhead compared to plain atomic reference counting. The design is described in the "Biased Reference Counting" section of PEP 703.

<!-- gh-issue-number: gh-110481 -->
* Issue: gh-110481
<!-- /gh-issue-number -->

colesbury committed:
Implement biased reference counting

colesbury committed:
Remove trailing commas in _PyObject_HEAD_INIT and _PyVarObject_HEAD_INIT

Removing the trailing comma allows the macros to be used in more places,
which means fewer `#ifdef Py_NOGIL` checks.

Note that the public API macros `PyObject_HEAD_INIT` and
`PyVarObject_HEAD_INIT` are not changed.

colesbury committed:
Add blurb

corona10 commented on Include/internal/pycore_object.h:
```diff
@@ -95,24 +103,63 @@ static inline void _Py_RefcntAdd(PyObject* op, Py_ssize_t n)
 {
     if (op) {
+#ifdef Py_NOGIL
+        op->ob_tid = 0;
```
Can we declare the constant variable for this? Something like `UNKNOWN_TID`?
I mean if someone already read the PEP, the person knows what the state that tid is zero stands for, but if not some kind of variable name will be helpful to read.

But I am not sure which will be the best naming

colesbury replied:
I've named it `_Py_UNOWNED_TID` because these are objects that are not owned by any thread.

corona10 commented on Include/object.h:
```diff
@@ -183,9 +226,56 @@ typedef struct {
+
+#if defined(Py_NOGIL) && !defined(Py_LIMITED_API)
+static inline Py_ALWAYS_INLINE int
+_Py_ThreadLocal(PyObject *ob)
```
nit: For a better understanding of what's happening.
Or you can use better naming to represent what this utility function is used for.
```suggestion
_Py_IsCurrentThreadOwn(PyObject *ob)
```

colesbury replied:
I've named it `_Py_IsOnwedByCurrentThread`, which is a bit long, but I think accurately describes what it checks.

erlend-aasland replied:
> I've named it `_Py_IsOnwedByCurrentThread`, which is a bit long, but I think accurately describes what it checks.

Did you mean `_Py_IsOwnedByCurrentThread`? :)

corona10 commented on Include/object.h:
```diff
@@ -643,6 +764,19 @@ static inline Py_ALWAYS_INLINE void Py_INCREF(PyObject *op)
+// zero. The call will deallocate the object if the shared refcount is also
+// zero. Otherwise, the thread gives up ownership and merges the reference
+// count fields.
+PyAPI_FUNC(void) _Py_MergeZeroRefcount(PyObject *);
```
When I read the code, the purpose of this function is to handle the object when the local ref count is zero.
So might be, we can improve function name to understand what's happening
```suggestion
PyAPI_FUNC(void) _Py_MergeZeroRefLocalcount(PyObject *);
```

or 

```suggestion
PyAPI_FUNC(void) _Py_MergeZeroLocalRefcount(PyObject *);
```

colesbury replied:
I've named it `_Py_MergeZeroLocalRefcount`

corona10 reviewed (approved):
Overall looks good to me.
I just leave the styling things for other readers.
IMO, It will help to understand the mechanism between local refcount and shared ref count.

corona10 commented on Objects/object.c:
```diff
@@ -295,6 +295,123 @@ _Py_DecRef(PyObject *o)
+Py_ssize_t
+_Py_ExplicitMergeRefcount(PyObject *op, Py_ssize_t extra)
+{
+    Py_ssize_t refcnt;
```
Out of curiosity, Don't we have to add assert statement that op should not be an immortal object?

````c
assert(!_Py_IsImmortal(op));
````

colesbury replied:
Yeah, that's a good assertion to add.

colesbury committed:
Changes from review:

 - _Py_ThreadLocal -> _Py_IsOnwedByCurrentThread
 - Add _Py_UNOWNED_TID macro define
 - _Py_MergeZeroRefcount -> Py_MergeZeroLocalRefcount
 - Add assertion to _Py_ExplicitMergeRefcount

colesbury committed:
Fix compile errors from rebasing on main

colesbury reviewed (commented):
Thanks for reviewing this! I've updated the PR and rebased on the main branch.

corona10 commented on Modules/_testcapi/heaptype_relative.c:
```diff
@@ -1,4 +1,11 @@
+#ifndef _MSC_VER
+#include "pyconfig.h"   // Py_NOGIL
```
Hmm IIUC, parts.h includes python.h and python.h includes pyconfig.h
I think that @vstinner or @erlend-aasland can left the comment for this.
https://github.com/python/cpython/blob/677d4bc15e999aa57abe7d23d7fac2c34a6fe0ba/Include/Python.h#L12-L15

Please let me know if I miss something @colesbury

erlend-aasland replied:
Yeah, but I also think it is fine to include it explicitly. OTOH, it would make the diff smaller , FWIW.

Why is it only included on Windows?

colesbury replied:
* We can't define `Py_LIMITED_API` in `--disable-gil` builds because the required atomic operations are not available in the limited API and `--disable-gil` builds are not compatible with the stable ABI. I expect this to change in the future -- there was a productive discussion at the sprint about this, but it'll require a PEP and take a while.
* So we need to guard the `#define Py_LIMITED_API 0x030...` with a `#ifndef Py_NOGIL` for now
* On `configure` based systems (Linux and macOS), `Py_NOGIL` is defined in `pyconfig.h`, so we need to include that before setting `Py_LIMITED_API`. We can't include all of `"Python.h"` because that definition depends on `Py_LIMITED_API`
* The situation is different on Windows. On Windows, `Py_NOGIL` is passed as a build flag and the Windows [`pyconfig.h`](https://github.com/python/cpython/blob/11312eae6ec3acf51aacafce4cb6d1a5edfd5f2e/PC/pyconfig.h#L312) depends on `Py_LIMITED_API` so we should *not* include it before setting `Py_LIMITED_API`.

erlend-aasland replied:
Ok, thanks for the details :)

corona10 reviewed (changes requested):
@colesbury In my macOS, the PR is failed to build :) Pleace check

````
9 warnings generated.
gcc  -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall    -std=c11 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration -fvisibility=hidden  -I./Include/internal  -I. -I./Include     -c ./Modules/xxsubtype.c -o Modules/xxsubtype.o
In file included from ./Modules/_scproxy.c:9:
In file included from ./Include/Python.h:52:
./Include/object.h:276:22: error: implicit declaration of function '_Py_atomic_load_uint32_relaxed' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    uint32_t local = _Py_atomic_load_uint32_relaxed(&ob->ob_ref_local);
                     ^
./Include/object.h:280:25: error: implicit declaration of function '_Py_atomic_load_ssize_relaxed' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    Py_ssize_t shared = _Py_atomic_load_ssize_relaxed(&ob->ob_ref_shared);
                        ^
./Include/object.h:280:25: note: did you mean '_Py_atomic_load_uint32_relaxed'?
./Include/object.h:276:22: note: '_Py_atomic_load_uint32_relaxed' declared here
    uint32_t local = _Py_atomic_load_uint32_relaxed(&ob->ob_ref_local);
                     ^
./Include/object.h:343:9: error: implicit declaration of function '_Py_IsOnwedByCurrentThread' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    if (_Py_IsOnwedByCurrentThread(ob)) {
        ^
````

colesbury committed:
Fix typo

colesbury committed:
Add missing  guards

vstinner commented on Include/internal/pycore_object.h:
```diff
@@ -54,16 +54,24 @@ PyAPI_FUNC(int) _PyObject_IsFreed(PyObject *);
+        .ob_base = _PyObject_HEAD_INIT(type), \
         .ob_size = size                       \
-    },
+    }
```
While I hate this comma, I would prefer that you not change it as part of this very complicated change. Would you mind to keep the comma to reduce the size of the PR?

colesbury replied:
I explained this in my first comment on this PR. It makes the PR smaller, not larger, because it avoids a number of `#ifdef Py_NOGIL` guards that would be necessary if the comma were still there. I put this as it's own commit (the first in the PR) so that the other changes can be reviewed separately.

vstinner replied:
Oh ok, sorry, I didn't see your comment. Now it makes sense to me.

vstinner commented on Include/object.h:
```diff
@@ -106,9 +106,26 @@ check by comparing the reference count field to the immortality reference count.
 // Make all internal uses of PyObject_HEAD_INIT immortal while preserving the
 // C-API expectation that the refcnt will be set to 1.
-#ifdef Py_BUILD_CORE
+#if defined(Py_NOGIL)
```
```suggestion
#ifdef Py_NOGIL
```

vstinner commented on Include/object.h:
```diff
@@ -183,9 +231,56 @@ typedef struct {
+#elif defined(__aarch64__)
+    __asm__ ("mrs %0, tpidr_el0" : "=r" (tid));
+#else
+  # error "define _Py_ThreadId for this platform"
```
```suggestion
#  error "define _Py_ThreadId for your platform"
```

vstinner commented on Include/object.h:
```diff
@@ -183,9 +231,56 @@ typedef struct {
 PyAPI_FUNC(int) Py_Is(PyObject *x, PyObject *y);
 #define Py_Is(x, y) ((x) == (y))
 
+#ifndef Py_LIMITED_API
```
Non-limited API must be in Include/cpython/object.h.

colesbury replied:
That's not possible here and `Include/object.h` already includes a number of non-limited API functions. It's not as cleanly separated as other file pairs. In particular, it's not possible to move this to `Include/cpython/object.h` because the definition is required by `Py_INCREF` and a number of other functions that must be defined before `Include/cpython/object.h` is included.

vstinner replied:
I wrote PR #111505 to fix the inter-dependency issue: move non-limited code to Include/cpython/object.h, and only keep limited code in Include/object.h.

vstinner reviewed (changes requested):
Sorry, this PR has too many issues, I cannot endorse it in its current shape.

I don't understand why _stat.c can no longer be built with the limited C API. For me, that's a no-go.

Please move all non-limited code to Include/cpython/object.h. I don't want to see assembly code in the limited C API. Maybe a first PR to move Py_INCREF/Py_DECREF to Include/cpython/ is needed.

colesbury reviewed (commented):
There is no assembly in the limited API. `_stat.c` is still built with the limited API C API in the default build. I explained the reasons why it's not currently possible to do so with the `--disable-gil` builds earlier. I'll copy-and-paste it below:
* We can't define `Py_LIMITED_API` in `--disable-gil` builds because the required atomic operations are not available in the limited API and `--disable-gil` builds are not compatible with the stable ABI. I expect this to change in the future -- there was a productive discussion at the sprint about this, but it'll require a PEP and take a while.
* So we need to guard the `#define Py_LIMITED_API 0x030...` with a `#ifndef Py_NOGIL` for now
* On `configure` based systems (Linux and macOS), `Py_NOGIL` is defined in `pyconfig.h`, so we need to include that before setting `Py_LIMITED_API`. We can't include all of `"Python.h"` because that definition depends on `Py_LIMITED_API`
* The situation is different on Windows. On Windows, `Py_NOGIL` is passed as a build flag and the Windows [`pyconfig.h`](https://github.com/python/cpython/blob/11312eae6ec3acf51aacafce4cb6d1a5edfd5f2e/PC/pyconfig.h#L312) depends on `Py_LIMITED_API` so we should *not* include it before setting `Py_LIMITED_API`.

corona10 reviewed (approved):
LGTM 

@colesbury 
Once you implement the inter-thread queue,
Would you like to measure pyperformance benchmark with GIL and noGIL version?
I would like to see how biased reference counting will impact to the single thread performance itself.

I leave the Py_INCREF / Py_DECREF part(about maintaining API) to @vstinner

markshannon commented on Include/internal/pycore_object.h:
```diff
@@ -54,16 +54,24 @@ PyAPI_FUNC(int) _PyObject_IsFreed(PyObject *);
         .ob_refcnt = _Py_IMMORTAL_REFCNT, \
         .ob_type = (type)                 \
-    },
+    }
```
Loosing the trailing comma is definitely better style, but is bound to break some third party code.
I think you need to leave the comma.

colesbury replied:
This is in an internal-only header (`Include/internal/pycore_object.h`), not used by third party code.

There's a similarly named public macro used by third party code that is not changed.

markshannon commented on Include/internal/pycore_object.h:
```diff
@@ -95,24 +103,63 @@ static inline void _Py_RefcntAdd(PyObject* op, Py_ssize_t n)
 #ifdef Py_REF_DEBUG
     _Py_AddRefTotal(_PyInterpreterState_GET(), n);
 #endif
+#if !defined(Py_NOGIL)
```
Is this worth the extra code?
`_Py_RefcntAdd` is only used in `list_repeat` and `tuplerepeat`. Multiplying tuples and lists by ints is hardly a common operation.
Would it perhaps make more sense to remove `_Py_RefcntAdd` and use plain `Py_INCREF` in those functions?

colesbury replied:
Are you suggesting removing `_Py_RefcntAdd` in general or just for `Py_NOGIL`?

markshannon commented on Include/internal/pycore_object.h:
```diff
@@ -161,6 +209,36 @@ _Py_DECREF_NO_DEALLOC(PyObject *op)
 
+#else
+static inline void
+_Py_DECREF_SPECIALIZED(PyObject *op, const destructor destruct)
```
Why not keep the specialization?

colesbury replied:
I'm trying to keep the PR small so that it's easier to review. I intend to implement the specializations later.

markshannon commented on Include/internal/pycore_object.h:
```diff
@@ -161,6 +209,36 @@ _Py_DECREF_NO_DEALLOC(PyObject *op)
+}
+
+static inline void
+_Py_DECREF_NO_DEALLOC(PyObject *op)
```
Given we know that the refcount cannot reach zero, why not take advantage of that?

colesbury replied:
I intend to implement this later.

markshannon commented on Include/internal/pycore_object.h:
```diff
@@ -95,24 +103,63 @@ static inline void _Py_RefcntAdd(PyObject* op, Py_ssize_t n)
 
+// Makes an immortal object mortal again with the specified refcnt. Should only
+// be used during runtime finalization.
+static inline void _Py_SetMortal(PyObject *op, Py_ssize_t refcnt)
```
This function shouldn't exist.
Making an immortal object mortal is not safe.
If it is used during runtime finalization, then the finalization should be fixed.

colesbury replied:
This is important for the reference leak checks. It's not new functionality. Without it, we would leak immortal objects on interpreter shutdown. It's not generally safe, which is why it's in an `Include/internal` header. 

We use this in two places. In both our intention is to deallocate immortal objects.

1. In `_PyUnicode_ClearInterned` https://github.com/python/cpython/blob/3726cb0f146cb229a5e9db8d41c713b023dcd474/Objects/unicodeobject.c#L14967-L14970
2. In `_Py_ClearImmortal` to deallocate immortal objects.

markshannon commented on Python/ceval.c:
```diff
@@ -46,7 +46,7 @@
 #endif
 
-#if !defined(Py_DEBUG) && !defined(Py_TRACE_REFS)
+#if !defined(Py_DEBUG) && !defined(Py_TRACE_REFS) && !defined(Py_NOGIL)
```
Have you checked what the performance impact of this on windows is?

colesbury replied:
It seems roughly an extra ~10% regression on Windows pgo. I intend to implement this later, but I'd like to keep this PR small.

colesbury reviewed (commented):
Hi @markshannon, thanks for reviewing this. This PR is a blocker for nearly all the subsequent work on nogil. It doesn't make sense to tackle deferred reference counting first for a few reasons. First, the implementation of deferred reference counting depends on the implementation of biased reference counting. Second, deferred reference counting is primarily an optimization for scaling multithreaded programs. It's not necessary for correctness and there are many correctness/thread-safety changes that need to be made. Additionally, deferred reference counting will not reduce the number of `#ifdef Py_NOGIL` guards -- it won't reduce any divergence introduced here.

colesbury committed:
Add TODOs for missing Py_DECREF specializations

These aren't necessary for correctness, just performance.

corona10 reviewed (approved):
@colesbury cc @vstinner @ambv 
LGTM Again

Let's not block the PEP-703 works for the things.
It will not affect the default implementation, and for the limited C API issue, I believe that we can solve it before we want to use `free-threaded` mode by default.

Let's solve it as a separate issue.

dimaqq commented on Include/object.h:
```diff
@@ -183,9 +231,56 @@ typedef struct {
+    __asm__("movq %%gs:0, %0" : "=r" (tid));  // x86_64 macOSX uses GS
+#elif defined(__x86_64__)
+   __asm__("movq %%fs:0, %0" : "=r" (tid));  // x86_64 Linux, BSD uses FS
+#elif defined(__arm__)
+    __asm__ ("mrc p15, 0, %0, c13, c0, 3\nbic %0, %0, #3" : "=r" (tid));
+#elif defined(__aarch64__) && defined(__APPLE__)
+    __asm__ ("mrs %0, tpidrro_el0" : "=r" (tid));
+#elif defined(__aarch64__)
+    __asm__ ("mrs %0, tpidr_el0" : "=r" (tid));
+#else
```
Caveat: I'm not an expert.
In case of x86/x86-64 above, the tid values has to be copied to a regular register for comparison.
In case of ARM32, the comparison could be made directly, were it not for the 2 lsb.
In case of ARM64, the comparison could be made directly.

Leaving aside the less common case of stamping the biased reference count block, and focusing on the more common case of comparing the current tid with object owner tid:

Does the compiler optimise this move away?
Does it perhaps not matter, as move is fast compared to loading the object's value from ram or cache?
Or would a custom comparison macro make sense here?

colesbury replied:
This is backwards: on x86/x86-64 you can do a comparison without copying `tid` to a register; you can't on ARM.

On x86, many instructions, such as `cmp`, can operate on memory addresses, including addresses with `fs:` or `gs:` prefixes. It doesn't matter much here because because the object's owner tid also requires a memory load. So either one or the other needs to be moved to a temporary registers and it's more convenient to move the current tid to a temporary register because it may be re-used across multiple reference count operations.

On ARM you can't do a comparison with a system or coprocessor register. You need to move it to a gpr first.
